File: backend/tools/run_command.py
import subprocess
from workspace import get_workspace

async def run_command(command: str, timeout: int = 30):
    workspace = get_workspace()
    if not workspace:
        raise ValueError("Workspace not set")

    result = subprocess.run(
        command,
        cwd=workspace,
        shell=True,
        capture_output=True,
        text=True,
        timeout=timeout
    )

    return {
        "command": command,
        "return_code": result.returncode,
        "stdout": result.stdout,
        "stderr": result.stderr
    }

File: frontend/src/tools/runCommand.ts
import type { Tool } from "./types"
import { callWorkspaceApi } from "./backendApi"

export const runCommandTool: Tool = {
  name: "run_command",
  description: "Run a shell command in the current workspace and return stdout/stderr.",
  parameters: {
    type: "object",
    properties: {
      command: { type: "string" },
      timeout: { type: "integer" }
    },
    required: ["command"]
  },
  async execute({ command, timeout = 30 }) {
    return await callWorkspaceApi("/workspace/command", { command, timeout })
  }
}
